src/Nethermind/Nethermind.HonestValidator/HonestValidatorWorker.cs

```diff
@@ -0,0 +1,131 @@
+            ClientVersion clientVersion)
+        {
+            _logger = logger;
+            _clock = clock;
+            _environment = environment;
+            _configuration = configuration;
+            _beaconNodeApi = beaconNodeApi;
+            _beaconChain = beaconChain;
+            _validatorClient = validatorClient;
+            _clientVersion = clientVersion;
+        }
+
+        public override async Task StopAsync(CancellationToken cancellationToken)
+        {
+            if (_logger.IsDebug()) LogDebug.HonestValidatorWorkerStopping(_logger, null);
+            _stopped = true;
+            await base.StopAsync(cancellationToken);
+        }
+
+        public override async Task StartAsync(CancellationToken cancellationToken)
+        {
+            if (_logger.IsDebug()) LogDebug.HonestValidatorWorkerStarting(_logger, null);
+            await base.StartAsync(cancellationToken);
+            if (_logger.IsDebug()) LogDebug.HonestValidatorWorkerStarted(_logger, null);
+        }
+
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            if (_logger.IsInfo())
+                Log.HonestValidatorWorkerExecuteStarted(_logger, _clientVersion.Description,
+                    _environment.EnvironmentName, Thread.CurrentThread.ManagedThreadId, null);
+
+            try
+            {
+                // Config
+                // List of nodes
+                // Validator private keys (or quickstart)
+                // Seconds per slot
+                
+                string nodeVersion = await _beaconNodeApi.GetNodeVersionAsync(stoppingToken).ConfigureAwait(false);
+                ulong genesisTime = await _beaconNodeApi.GetGenesisTimeAsync(stoppingToken).ConfigureAwait(false);
+                Log.HonestValidatorWorkerConnected(_logger, nodeVersion, genesisTime, null);
+                
+                await _beaconChain.SetGenesisTimeAsync(genesisTime).ConfigureAwait(false);
+                
+                while (!stoppingToken.IsCancellationRequested && !_stopped)
+                {
+                    try
+                    {
+                        DateTimeOffset clockTime = _clock.UtcNow();
+                        ulong time = (ulong) clockTime.ToUnixTimeSeconds();
+
+                        await _validatorClient.OnTickAsync(_beaconChain, time, stoppingToken).ConfigureAwait(false);
+                        
+                        // Wait for remaining time, if any
+                        // NOTE: To fast forward time during testing, have the second call to test _clock.Now() jump forward to avoid waiting.
+                        DateTimeOffset nextClockTime = DateTimeOffset.FromUnixTimeSeconds((long) time + 1);
+                        TimeSpan remaining = nextClockTime - _clock.UtcNow();
+                        if (remaining > TimeSpan.Zero)
+                        {
+                            await Task.Delay(remaining, stoppingToken);
```

Comment:
is this better than System.Timer?

Reply:
Hmm. So, they would pretty much do the same thing, except System.Timer doesn't support async directly. You could have it calling async void, but then would have to make sure to handle any exceptions properly. 

Internally Task.Delay uses Timer, just seems to have neater exception handling.

The other difference is if the processing takes a long time. Often Task/Delay has a fixed delay (e.g. 1 sec) between calls, whereas Timer triggers a call every second, even if the first hasn't completed.

However, rather than a fixed delay I work out how long until the next second boundary, so it works similar to Timer.

...Except where processing takes longer than 1 sec, in which case it will wait, trigger immediately (no delay), but only trigger once (e.g. if took > 2 secs).

(You could, of course, emulate this behaviour with Timer, but async seems pretty neat)

Reply:
Just a siede note Task.Delay has terrible resolution. It can return before the actual delay will happen(!). I had this issue with AuRa when I need to have at least some delay. I created TaskExt.DelayAtLeast method for that. Not sure if this would case issue for you.

https://stackoverflow.com/questions/47949709/task-delay-completes-earlier